File: src/claw_mem/links/memory_links.py
# Copyright 2026 Peter Cheng
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
claw-mem Memory Links and Tags

Simple markdown-based linking and tagging system.
Maintains simplicity while adding association capabilities.

Syntax:
- Links: [[memory_id]] or [[2026-03-25#investment决策]]
- Tags: #tag or #tagname
"""

import re
from typing import Dict, List, Optional, Set, Tuple
from dataclasses import dataclass, field
from datetime import datetime


@dataclass
class MemoryLink:
    """Memory link"""
    source_id: str
    target_id: str
    link_type: str = "reference"  # reference, related, see_also
    created_at: str = field(default_factory=lambda: datetime.now().isoformat())
    
    def to_dict(self) -> Dict:
        return {
            "source_id": self.source_id,
            "target_id": self.target_id,
            "link_type": self.link_type,
            "created_at": self.created_at
        }


@dataclass
class MemoryTags:
    """Memory tags"""
    memory_id: str
    tags: Set[str] = field(default_factory=set)
    updated_at: str = field(default_factory=lambda: datetime.now().isoformat())
    
    def to_dict(self) -> Dict:
        return {
            "memory_id": self.memory_id,
            "tags": list(self.tags),
            "updated_at": self.updated_at
        }


class MemoryLinkParser:
    """
    Memory Link Parser
    
    Parses [[memory_id]] syntax from markdown content.
    """
    
    # Pattern: [[memory_id]] or [[date#title]]
    LINK_PATTERN = re.compile(r'\[\[([^\]]+)\]\]')
    
    def parse_links(self, content: str, source_id: str) -> List[MemoryLink]:
        """
        Parse links from content
        
        Args:
            content: Markdown content
            source_id: Source memory ID
            
        Returns:
            List[MemoryLink]: Parsed links
        """
        links = []
        
        for match in self.LINK_PATTERN.finditer(content):
            target = match.group(1).strip()
            
            # Parse target (could be memory_id or date#title)
            if '#' in target:
                # date#title format
                date_part, title_part = target.split('#', 1)
                target_id = f"{date_part}#{title_part}"
            else:
                # Simple memory_id
                target_id = target
            
            links.append(MemoryLink(
                source_id=source_id,
                target_id=target_id,
                link_type="reference"
            ))
        
        return links
    
    def remove_links(self, content: str) -> str:
        """
        Remove link syntax from content
        
        Args:
            content: Markdown content
            
        Returns:
            str: Content without link syntax
        """
        # Replace [[memory_id]] with memory_id
        return self.LINK_PATTERN.sub(r'\1', content)
    
    def extract_link_targets(self, content: str) -> List[str]:
        """
        Extract link targets from content
        
        Args:
            content: Markdown content
            
        Returns:
            List[str]: Link targets
        """
        return [
            match.group(1).strip()
            for match in self.LINK_PATTERN.finditer(content)
        ]


class MemoryTagParser:
    """
    Memory Tag Parser
    
    Parses #tag syntax from markdown content.
    """
    
    # Pattern: #tag or #tagname (Unicode supported)
    TAG_PATTERN = re.compile(r'#([a-zA-Z0-9\u4e00-\u9fa5_]+)')
    
    def parse_tags(self, content: str) -> Set[str]:
        """
        Parse tags from content
        
        Args:
            content: Markdown content
            
        Returns:
            Set[str]: Parsed tags
        """
        tags = set()
        
        for match in self.TAG_PATTERN.finditer(content):
            tag = match.group(1).strip()
            tags.add(tag.lower())  # Normalize to lowercase
        
        return tags
    
    def remove_tags(self, content: str) -> str:
        """
        Remove tag syntax from content
        
        Args:
            content: Markdown content
            
        Returns:
            str: Content without tag syntax
        """
        # Replace #tag with empty string
        return self.TAG_PATTERN.sub('', content)
    
    def extract_tags(self, content: str) -> Set[str]:
        """
        Extract tags from content
        
        Args:
            content: Markdown content
            
        Returns:
            Set[str]: Tags
        """
        return self.parse_tags(content)


class MemoryLinkManager:
    """
    Memory Link Manager
    
    Manages memory links and tags.
    """
    
    def __init__(self):
        """Initialize Link Manager"""
        self.link_parser = MemoryLinkParser()
        self.tag_parser = MemoryTagParser()
        
        # In-memory storage
        self.links: Dict[str, List[MemoryLink]] = {}  # source_id -> links
        self.tags: Dict[str, MemoryTags] = {}  # memory_id -> tags
        
        # Reverse index for tag search
        self.tag_index: Dict[str, Set[str]] = {}  # tag -> memory_ids
    
    def process_memory(
        self,
        memory_id: str,
        content: str
    ) -> Tuple[List[MemoryLink], Set[str]]:
        """
        Process a memory (extract links and tags)
        
        Args:
            memory_id: Memory ID
            content: Memory content
            
        Returns:
            Tuple: (links, tags)
        """
        # Extract links
        links = self.link_parser.parse_links(content, memory_id)
        
        # Extract tags
        tags = self.tag_parser.parse_tags(content)
        
        # Store links
        self.links[memory_id] = links
        
        # Store tags
        self._update_tags(memory_id, tags)
        
        return links, tags
    
    def _update_tags(self, memory_id: str, tags: Set[str]) -> None:
        """Update tags for a memory"""
        if memory_id in self.tags:
            # Remove from old tag index
            old_tags = self.tags[memory_id].tags
            for tag in old_tags:
                if tag in self.tag_index:
                    self.tag_index[tag].discard(memory_id)
        
        # Update tags
        self.tags[memory_id] = MemoryTags(
            memory_id=memory_id,
            tags=tags,
            updated_at=datetime.now().isoformat()
        )
        
        # Update tag index
        for tag in tags:
            if tag not in self.tag_index:
                self.tag_index[tag] = set()
            self.tag_index[tag].add(memory_id)
    
    def get_linked_memories(self, memory_id: str) -> List[str]:
        """
        Get memories linked from a memory
        
        Args:
            memory_id: Memory ID
            
        Returns:
            List[str]: Linked memory IDs
        """
        if memory_id not in self.links:
            return []
        
        return [link.target_id for link in self.links[memory_id]]
    
    def get_backlinks(self, memory_id: str) -> List[str]:
        """
        Get memories that link to this memory
        
        Args:
            memory_id: Memory ID
            
        Returns:
            List[str]: Backlink memory IDs
        """
        backlinks = []
        
        for source_id, links in self.links.items():
            for link in links:
                if link.target_id == memory_id:
                    backlinks.append(source_id)
        
        return backlinks
    
    def search_by_tag(self, tag: str) -> List[str]:
        """
        Search memories by tag
        
        Args:
            tag: Tag name
            
        Returns:
            List[str]: Memory IDs with this tag
        """
        tag_lower = tag.lower()
        return list(self.tag_index.get(tag_lower, set()))
    
    def get_tags_for_memory(self, memory_id: str) -> Set[str]:
        """
        Get tags for a memory
        
        Args:
            memory_id: Memory ID
            
        Returns:
            Set[str]: Tags
        """
        if memory_id not in self.tags:
            return set()
        
        return self.tags[memory_id].tags.copy()
    
    def get_all_tags(self) -> Set[str]:
        """
        Get all tags
        
        Returns:
            Set[str]: All tags
        """
        return set(self.tag_index.keys())
    
    def get_related_memories(
        self,
        memory_id: str,
        limit: int = 10
    ) -> List[str]:
        """
        Get related memories (links + shared tags)
        
        Args:
            memory_id: Memory ID
            limit: Result limit
            
        Returns:
            List[str]: Related memory IDs
        """
        related = set()
        
        # Add linked memories
        linked = self.get_linked_memories(memory_id)
        related.update(linked)
        
        # Add backlinks
        backlinks = self.get_backlinks(memory_id)
        related.update(backlinks)
        
        # Add memories with shared tags
        my_tags = self.get_tags_for_memory(memory_id)
        for tag in my_tags:
            tagged_memories = self.search_by_tag(tag)
            related.update(tagged_memories)
        
        # Remove self
        related.discard(memory_id)
        
        # Return limited list
        return list(related)[:limit]
    
    def export_links(self) -> Dict:
        """
        Export links and tags
        
        Returns:
            Dict: Exported data
        """
        return {
            "links": {
                source: [link.to_dict() for link in links]
                for source, links in self.links.items()
            },
            "tags": {
                memory_id: tags.to_dict()
                for memory_id, tags in self.tags.items()
            },
            "tag_index": {
                tag: list(memories)
                for tag, memories in self.tag_index.items()
            }
        }

File: src/links/memory_links.ts
// Copyright 2026 Peter Cheng
// Licensed under Apache-2.0

/**
 * claw-mem Memory Links and Tags
 *
 * Simple markdown-based linking and tagging system.
 * Maintains simplicity while adding association capabilities.
 *
 * Syntax:
 * - Links: [[memory_id]] or [[2026-03-25#investment-decision]]
 * - Tags: #tag or #tagname
 */

export interface MemoryLinkData {
  source_id: string;
  target_id: string;
  link_type: string;
  created_at: string;
}

export class MemoryLink {
  sourceId: string;
  targetId: string;
  linkType: string;
  createdAt: string;

  constructor(
    sourceId: string,
    targetId: string,
    linkType: string = "reference",
    createdAt?: string,
  ) {
    this.sourceId = sourceId;
    this.targetId = targetId;
    this.linkType = linkType;
    this.createdAt = createdAt ?? new Date().toISOString();
  }

  toDict(): MemoryLinkData {
    return {
      source_id: this.sourceId,
      target_id: this.targetId,
      link_type: this.linkType,
      created_at: this.createdAt,
    };
  }
}

export interface MemoryTagsData {
  memory_id: string;
  tags: string[];
  updated_at: string;
}

export class MemoryTags {
  memoryId: string;
  tags: Set<string>;
  updatedAt: string;

  constructor(memoryId: string, tags?: Set<string>, updatedAt?: string) {
    this.memoryId = memoryId;
    this.tags = tags ?? new Set();
    this.updatedAt = updatedAt ?? new Date().toISOString();
  }

  toDict(): MemoryTagsData {
    return {
      memory_id: this.memoryId,
      tags: [...this.tags],
      updated_at: this.updatedAt,
    };
  }
}

/**
 * Memory Link Parser
 *
 * Parses [[memory_id]] syntax from markdown content.
 */
export class MemoryLinkParser {
  // Pattern: [[memory_id]] or [[date#title]]
  private static LINK_RE = /\[\[([^\]]+)\]\]/g;

  /**
   * Parse links from content.
   *
   * @param content  - Markdown content
   * @param sourceId - Source memory ID
   * @returns Parsed MemoryLink instances
   */
  parseLinks(content: string, sourceId: string): MemoryLink[] {
    const links: MemoryLink[] = [];
    let match: RegExpExecArray | null;

    // Reset regex state
    MemoryLinkParser.LINK_RE.lastIndex = 0;

    while ((match = MemoryLinkParser.LINK_RE.exec(content)) !== null) {
      const target = match[1].trim();

      // Parse target (could be memory_id or date#title)
      let targetId: string;
      if (target.includes("#")) {
        const [datePart, titlePart] = target.split("#", 2);
        targetId = `${datePart}#${titlePart}`;
      } else {
        targetId = target;
      }

      links.push(new MemoryLink(sourceId, targetId, "reference"));
    }

    return links;
  }

  /**
   * Remove link syntax from content, leaving just the link text.
   *
   * @param content - Markdown content
   * @returns Content without link syntax
   */
  removeLinks(content: string): string {
    return content.replace(MemoryLinkParser.LINK_RE, "$1");
  }

  /**
   * Extract link targets from content.
   *
   * @param content - Markdown content
   * @returns List of link target strings
   */
  extractLinkTargets(content: string): string[] {
    const targets: string[] = [];
    let match: RegExpExecArray | null;

    MemoryLinkParser.LINK_RE.lastIndex = 0;

    while ((match = MemoryLinkParser.LINK_RE.exec(content)) !== null) {
      targets.push(match[1].trim());
    }

    return targets;
  }
}

/**
 * Memory Tag Parser
 *
 * Parses #tag syntax from markdown content.
 */
export class MemoryTagParser {
  // Pattern: #tag or #tagname (Unicode supported)
  private static TAG_RE = /#([a-zA-Z0-9\u4e00-\u9fa5_]+)/g;

  /**
   * Parse tags from content.
   *
   * @param content - Markdown content
   * @returns Set of parsed tags (lowercased)
   */
  parseTags(content: string): Set<string> {
    const tags = new Set<string>();
    let match: RegExpExecArray | null;

    MemoryTagParser.TAG_RE.lastIndex = 0;

    while ((match = MemoryTagParser.TAG_RE.exec(content)) !== null) {
      const tag = match[1].trim();
      tags.add(tag.toLowerCase());
    }

    return tags;
  }

  /**
   * Remove tag syntax from content.
   *
   * @param content - Markdown content
   * @returns Content without tag syntax
   */
  removeTags(content: string): string {
    return content.replace(MemoryTagParser.TAG_RE, "");
  }

  /**
   * Extract tags from content.
   *
   * @param content - Markdown content
   * @returns Set of tags
   */
  extractTags(content: string): Set<string> {
    return this.parseTags(content);
  }
}

/**
 * Memory Link Manager
 *
 * Manages memory links and tags with in-memory storage.
 */
export class MemoryLinkManager {
  linkParser: MemoryLinkParser;
  tagParser: MemoryTagParser;

  // source_id -> links
  private links: Map<string, MemoryLink[]>;
  // memory_id -> tags
  private tags: Map<string, MemoryTags>;
  // tag -> memory_ids
  private tagIndex: Map<string, Set<string>>;

  constructor() {
    this.linkParser = new MemoryLinkParser();
    this.tagParser = new MemoryTagParser();
    this.links = new Map();
    this.tags = new Map();
    this.tagIndex = new Map();
  }

  /**
   * Process a memory: extract links and tags from content.
   *
   * @param memoryId - Memory ID
   * @param content  - Memory content
   * @returns Tuple of [links, tags]
   */
  processMemory(memoryId: string, content: string): [MemoryLink[], Set<string>] {
    const extractedLinks = this.linkParser.parseLinks(content, memoryId);
    const extractedTags = this.tagParser.parseTags(content);

    // Store links
    this.links.set(memoryId, extractedLinks);

    // Store tags
    this._updateTags(memoryId, extractedTags);

    return [extractedLinks, extractedTags];
  }

  private _updateTags(memoryId: string, newTags: Set<string>): void {
    // Remove from old tag index
    if (this.tags.has(memoryId)) {
      const oldTags = this.tags.get(memoryId)!.tags;
      for (const tag of oldTags) {
        const idx = this.tagIndex.get(tag);
        if (idx) {
          idx.delete(memoryId);
          if (idx.size === 0) {
            this.tagIndex.delete(tag);
          }
        }
      }
    }

    // Update tags
    this.tags.set(
      memoryId,
      new MemoryTags(memoryId, newTags, new Date().toISOString()),
    );

    // Update tag index
    for (const tag of newTags) {
      if (!this.tagIndex.has(tag)) {
        this.tagIndex.set(tag, new Set());
      }
      this.tagIndex.get(tag)!.add(memoryId);
    }
  }

  /**
   * Get memory IDs linked from a memory.
   *
   * @param memoryId - Memory ID
   * @returns List of linked memory IDs
   */
  getLinkedMemories(memoryId: string): string[] {
    const links = this.links.get(memoryId);
    if (!links) return [];
    return links.map((l) => l.targetId);
  }

  /**
   * Get memory IDs that link to this memory (backlinks).
   *
   * @param memoryId - Memory ID
   * @returns List of backlink memory IDs
   */
  getBacklinks(memoryId: string): string[] {
    const backlinks: string[] = [];
    for (const [sourceId, linkList] of this.links) {
      for (const link of linkList) {
        if (link.targetId === memoryId) {
          backlinks.push(sourceId);
        }
      }
    }
    return backlinks;
  }

  /**
   * Search memories by tag.
   *
   * @param tag - Tag name
   * @returns Memory IDs with this tag
   */
  searchByTag(tag: string): string[] {
    const lower = tag.toLowerCase();
    return [...(this.tagIndex.get(lower) ?? new Set())];
  }

  /**
   * Get tags for a memory.
   *
   * @param memoryId - Memory ID
   * @returns Set of tags
   */
  getTagsForMemory(memoryId: string): Set<string> {
    const entry = this.tags.get(memoryId);
    return entry ? new Set(entry.tags) : new Set();
  }

  /**
   * Get all tags in the system.
   *
   * @returns Set of all tag names
   */
  getAllTags(): Set<string> {
    return new Set(this.tagIndex.keys());
  }

  /**
   * Get related memories (links + shared tags).
   *
   * @param memoryId - Memory ID
   * @param limit    - Maximum results (default 10)
   * @returns Related memory IDs
   */
  getRelatedMemories(memoryId: string, limit: number = 10): string[] {
    const related = new Set<string>();

    // Add linked memories
    const linked = this.getLinkedMemories(memoryId);
    for (const id of linked) related.add(id);

    // Add backlinks
    const backlinks = this.getBacklinks(memoryId);
    for (const id of backlinks) related.add(id);

    // Add memories with shared tags
    const myTags = this.getTagsForMemory(memoryId);
    for (const tag of myTags) {
      const tagged = this.searchByTag(tag);
      for (const id of tagged) related.add(id);
    }

    // Remove self
    related.delete(memoryId);

    return [...related].slice(0, limit);
  }

  /**
   * Export all links and tags.
   *
   * @returns Exported data object
   */
  exportLinks(): Record<string, unknown> {
    const linksObj: Record<string, MemoryLinkData[]> = {};
    for (const [source, linkList] of this.links) {
      linksObj[source] = linkList.map((l) => l.toDict());
    }

    const tagsObj: Record<string, MemoryTagsData> = {};
    for (const [memId, tagEntry] of this.tags) {
      tagsObj[memId] = tagEntry.toDict();
    }

    const tagIdxObj: Record<string, string[]> = {};
    for (const [tag, memIds] of this.tagIndex) {
      tagIdxObj[tag] = [...memIds];
    }

    return {
      links: linksObj,
      tags: tagsObj,
      tag_index: tagIdxObj,
    };
  }
}
